File: Queue/Queue.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 29 18:12:54 2019

@author: musahni
"""

class Node(object):
    def __init__(self, item):
        self.item = item
        self.next = None
        
class Queue(object):
    def __init__(self):
        self.first = None
        self.last = None
        
    def isEmpty(self):
        return self.first == None
    
    def enqueue(self, node):
        old_last = self.last
        self.last = node
        self.last.next = None
        if(self.isEmpty() == True):
            self.first = self.last
        else:
            old_last.next = self.last
            
    def dequeue(self):
        item = self.first.item
        self.first = self.first.next
        
        if(self.isEmpty() == True):
            self.last = None
        
        return item
    
    
X = Node("X")
Y = Node("Y")

Q = Queue()

print(Q.isEmpty())
Q.enqueue(X)
Q.enqueue(Y)
print(Q.isEmpty())
print(Q.first.item)
print(Q.last.item)

File: _Queue/Queue.js
class Node {
    constructor(item) {
        this.item = item;
        this.next = null;
    };
};

class Queue {
    constructor() {
        this.first = null;
        this.last = null;
    };

    isEmpty() {
        return this.first == null;
    };

    enqueue(node) {
        let old_last = this.last;
        this.last = node;
        this.last.next = null;
        
        if(this.isEmpty() == true) {
            this.first = this.last;
        }
        else {
            old_last.next = this.last;
        };
    };

    dequeue() {
        item = this.first.item;
        this.first = this.first.next;

        if(this.isEmpty() == true) {
            this.last = null;
        };

        return item;
    };
};

let X = new Node("X");
let Y = new Node("Y");

let Q = new Queue();
console.log(`Is Q empty?: ${Q.isEmpty()}`);
console.log("Append to Q");
Q.enqueue(X);
console.log(`Is Q empty?: ${Q.isEmpty()}`);
console.log("Append to Q");
Q.enqueue(Y);
console.log(`First Q Node: ${Q.first.item}`);
console.log(`Last Q Node: ${Q.last.item}`);
